### ISSUE TITLE
bug/enhancement - latest_eventcount_5m from TrackMe metrics should perform an aggregation to properly represent the 5m sum of eventcounts

### ISSUE THREAD

**Issue by sebwurl** [trackme-limited/trackme-report-issues#94]
I don't know whether this really is a bug or some logic changed:
I Upgraded from 2.0.14 -> 2.0.19 on April 13th. Since that time, the trend for metrics changed significantly in the data source overview. Is this intended?
![image](https://user-images.githubusercontent.com/85566586/232755118-5230167f-5f1d-4d64-95af-07178d9597c0.png)

![image](https://user-images.githubusercontent.com/85566586/232756437-d79a4e72-0e2b-424c-b058-d406b211d830.png)

**Comment by sebwurl** [comment]
One further question regarding the mentioned part:

`| eval spantime=data_last_ingest | eventstats max(data_last_time_seen) as data_last_time_seen by index,sourcetype,host | eval spantime=if(spantime>=(now()-300), spantime, null())
| eventstats sum(data_eventcount) as eventcount_5m, avg(data_last_ingestion_lag_seen) as latency_5m, avg(dcount_host) as dcount_host_5m by spantime,index,sourcetype,host`

This adds a spantime for the 5-6 most current rows (in terms of the timestamp). The spantime is no bucket, so it is different for every of these lines. The following eventstats does not change anything because spantime are unique values. As a result, the last stats command just uses the eventcount of the most current row. This explains why the metrics event_count differs so significantly from the eventcount provided by the SPL search.

**Comment by guilhemmarchand** [comment]
@sebwurl 

```
Edit: Is the split by host in these lines added in the last upgrades? In my version only the tstats command is grouped by the host.
```

In latest releases, while creating an hybrid tracker:

<img width="1417" alt="image" src="https://user-images.githubusercontent.com/7503059/233137801-aebd7eb0-a1a5-4116-bb6c-849ae451b0e4.png">

So you can choose to include or not in the root breakby the splunk_server and host, which would add more accuracy in the latency calculation when using tstats, at the price of more cmputes.
In large scale environment where perf is a serious concern, it can be valuable to set these to False, you have a bit less of accuracy but still more than enough to detect trends, and use less computes so you end up with much faster trackers.

**Comment by guilhemmarchand** [comment]
Will come back on the previous question shortly

**Comment by sebwurl** [comment]
Right, I forgot about these new options. However, they don't have an effect on later commands of the search string.
Having break by splunk_server and host activated, it only effects the group by in the tstats command. In later commands the group by is just by sourcetype and index.

**Comment by guilhemmarchand** [comment]
@sebwurl 

`Right, I forgot about these new options. However, they don't have an effect on later commands of the search string.
Having break by splunk_server and host activated, it only effects the group by in the tstats command. In later commands the group by is just by sourcetype and index.`

No, it's incorrect.

Let's take the first part:

```
| tstats max(_indextime) as data_last_ingest, min(_time) as data_first_time_seen, max(_time) as data_last_time_seen, count as data_eventcount where index=eventgen-firewall-latency _index_earliest="-4h" _index_latest="+4h" by _time,index,sourcetype,splunk_server,host span=1s
| eval data_last_ingestion_lag_seen=data_last_ingest-data_last_time_seen
```

--> if you break by including the splunk_server, then you take into account per second one measure per indexer, in a clustered distributed context each indexer will receive potentially a subset of the events of the index/sourcetype/host, therefore you have more measures that you will take into account in the next aggregation segment

--> Same if you include the host, if these host are real endpoints (UF, etc) then each can have its own true latency, therefore you have more measures for the next aggregation

--> As with tstats we can't do the exat same calculation per single event (opposed to a raw search), then we take the max indextime and the max event timestamp and we calculate the latency

--> This remains approximative as on per second basis you have had more events than a single event, therefore the more you can include statements in the break by, the more you next aggregation level will be accurate, at the cost of having more measures to be taken into account and more computing costs in doing this task

The next part:

```
| bucket _time span=1m
| stats avg(data_last_ingestion_lag_seen) as data_last_ingestion_lag_seen, max(data_last_ingest) as data_last_ingest, min(data_first_time_seen) as data_first_time_seen, max(data_last_time_seen) as data_last_time_seen, sum(data_eventcount) as data_eventcount, dc(host) as dcount_host by _time,index,sourcetype
```

Aggregates the results on a per minute basis, on a per index/sourcetype (plus indexed fields if you have included a custom list of indexed fields.

The next part:

`| eval spantime=data_last_ingest | eventstats max(data_last_time_seen) as data_last_time_seen by index,sourcetype | eval spantime=if(spantime>=(now()-300), spantime, null())`

We define a pseudo time field against the last ingest, then we use a first eventstats to keep the latest _time (timestamp) per index/sourcetype, and finally we make sure that we are going to be taken into account only the latest 5 minutes of ingested data - this is useful to make sure we are not going to continue generating metrics if the data flow stops given that the global time range scope of the search is larger than that (and potentially very large)

The next part:

`| eventstats sum(data_eventcount) as eventcount_5m, avg(data_last_ingestion_lag_seen) as latency_5m, avg(dcount_host) as dcount_host_5m by spantime,index,sourcetype`

Aggregates the results using eventstats and include in the by statement the spantime, as we mare sure ealier that the spantime will be null expect for the last 5 minutes of ingest, we only take the last 5 minutes of ingested data without affecting the next part of the stream.

The next part:

```
| stats latest(eventcount_5m) as latest_eventcount_5m, avg(eventcount_5m) as avg_eventcount_5m, stdev(eventcount_5m) as stdev_eventcount_5m, perc95(eventcount_5m) as perc95_eventcount_5m, latest(latency_5m) as latest_latency_5m, avg(latency_5m) as avg_latency_5m, stdev(latency_5m) as stdev_latency_5m, perc95(latency_5m) as perc95_latency_5m, latest(dcount_host_5m) as latest_dcount_host_5m, avg(dcount_host_5m) as avg_dcount_host_5m, stdev(dcount_host_5m) as stdev_dcount_host_5m, perc95(dcount_host_5m) as perc95_dcount_host_5m, max(data_last_ingest) as data_last_ingest, min(data_first_time_seen) as data_first_time_seen, max(data_last_time_seen) as data_last_time_seen, avg(data_last_ingestion_lag_seen) as data_last_ingestion_lag_seen, sum(data_eventcount) as data_eventcount by index,sourcetype | eval dcount_host=round(latest_dcount_host_5m, 2)
 | eval data_last_ingestion_lag_seen=round(data_last_ingestion_lag_seen, 0) | rename index as data_index, sourcetype as data_sourcetype | eval object=data_index . ":" . data_sourcetype
``` tenant_id ```
| eval tenant_id="feeds-tracking"
``` call the abstract macro ```
`trackme_dsm_tracker_abstract(feeds-tracking, tstats)`
```

performs the final aggregation and submit the stream results to the abstract macro.

The metrics we generte are therefore an interpretation of the behaviour of the entity, it is not the proper eventcount for instance but the evencount movement over time from the ingest perspective, etc.
And it's important that we stop generating metrics when the flow itself stops.

**Comment by sebwurl** [comment]
I was confused because you mentioned this earlier:

_The important part is here:_

_| eval spantime=data_last_ingest | eventstats max(data_last_time_seen) as data_last_time_seen by index,sourcetype,**host** | eval spantime=if(spantime>=(now()-300), spantime, null())
| eventstats sum(data_eventcount) as eventcount_5m, avg(data_last_ingestion_lag_seen) as latency_5m, avg(dcount_host) as dcount_host_5m by spantime,index,sourcetype,**host**_

But the host does not appear in your last explanation (this is how it looks like in my environment as well).


---------

_The next part:_

_| eventstats sum(data_eventcount) as eventcount_5m, avg(data_last_ingestion_lag_seen) as latency_5m, avg(dcount_host) as dcount_host_5m by spantime,index,sourcetype_

_Aggregates the results using eventstats and include in the by statement the spantime, as we mare sure ealier that the spantime will be null expect for the last 5 minutes of ingest, we only take the last 5 minutes of ingested data without affecting the next part of the stream._

This comes back to my initial question, whether is intended **not** to use the sum event_count for all rows where the spantime is set. Because the spantime is unique for all rows, there is no aggregation.
I hope I am not misinterpreting it completely.

**Comment by guilhemmarchand** [comment]
`This comes back to my initial question, whether is intended not to use the sum event_count for all rows where the spantime is set. Because the spantime is unique for all rows, there is no aggregation.`

So, each row is a result for a given "entity" right. (index/sourcetype for instance)
When the spantime is set this means that we are looking at the last 5 minutes events that were ingested compared to now against the latest ingest date for that entity (now when the tracker runs)

So it does a sum of these events count.

So, "there is no aggregation" right, the point is to get how events were indexed in the past 5 minutes for each entity when the tracker runs, then turn this into multiple metrics (eventcount_5m, etc)

A simplistic example:

<img width="1774" alt="image" src="https://user-images.githubusercontent.com/7503059/233382508-9cefb7f3-9f3b-4aa3-bf30-451f50be6d86.png">

**Comment by sebwurl** [comment]
I agree that the final result is an aggregation. Here what I was expecting, I tried to visualize it somehow:
![expected value](https://user-images.githubusercontent.com/85566586/233454931-329f15f3-1839-423c-90bd-12dc90bbcf5f.png)

**Comment by guilhemmarchand** [comment]
Ok @sebwurl I am with you now ;-)

In fact, you are right - there is an issue in the logic with the event count, and most specially the latest_eventcount_5m which purposes was to show the sum of the latest eventcount per 5 min of data ingested in the entity.

Thefore:

<img width="1776" alt="image" src="https://user-images.githubusercontent.com/7503059/233469647-b78fe7c3-a438-4b33-9b3a-2bcdc9bab144.png">

Should end up with:

```
| stats sum(eventcount_5m) as latest_eventcount_5m, avg(eventcount_5m) as avg_eventcount_5m, stdev(eventcount_5m) as stdev_eventcount_5m, perc95(eventcount_5m) as perc95_eventcount_5m, latest(latency_5m) as latest_latency_5m, avg(latency_5m) as avg_latency_5m, stdev(latency_5m) as stdev_latency_5m, perc95(latency_5m) as perc95_latency_5m, latest(dcount_host_5m) as latest_dcount_host_5m, avg(dcount_host_5m) as avg_dcount_host_5m, stdev(dcount_host_5m) as stdev_dcount_host_5m, perc95(dcount_host_5m) as perc95_dcount_host_5m, max(data_last_ingest) as data_last_ingest, min(data_first_time_seen) as data_first_time_seen, max(data_last_time_seen) as data_last_time_seen, avg(data_last_ingestion_lag_seen) as data_last_ingestion_lag_seen, sum(data_eventcount) as data_eventcount by index,sourcetype | eval dcount_host=round(latest_dcount_host_5m, 2)
```

Where we do a "sum(eventcount_5m) as latest_eventcount_5m" instead of the latest to arrive to the wanted result (so the sum of all per min for the past 5 min)

The rest of the metrics is "ok" as they basically takeinto account the "variation" of every min of data available in the latest past 5 min.

But latest should have shown the sum because the point is to show this information when you open the entity.

So my apologises, you are right, there is an issue. (with this metric in particular)

<img width="1784" alt="image" src="https://user-images.githubusercontent.com/7503059/233470304-a09d86c5-40b9-41c8-aef1-b05920b5784e.png">


The fact too, for the ML side of thing, the default model is to generate a model against the avg eventcount per 5 min - I am wondering if given the latest evolutions it wouldn't make sense to do that against the latest instead once it is fixed.

So in anyway, do you agree?

I would fix this:

```
| stats latest(eventcount_5m) as latest_eventcount_5m, avg(eventcount_5m) as avg_eventcount_5m, stdev(eventcount_5m) as stdev_eventcount_5m, perc95(eventcount_5m) as perc95_eventcount_5m, latest(latency_5m) as latest_latency_5m, avg(latency_5m) as avg_latency_5m, stdev(latency_5m) as stdev_latency_5m, perc95(latency_5m) as perc95_latency_5m, latest(dcount_host_5m) as latest_dcount_host_5m, avg(dcount_host_5m) as avg_dcount_host_5m, stdev(dcount_host_5m) as stdev_dcount_host_5m, perc95(dcount_host_5m) as perc95_dcount_host_5m, max(data_last_ingest) as data_last_ingest, min(data_first_time_seen) as data_first_time_seen, max(data_last_time_seen) as data_last_time_seen, avg(data_last_ingestion_lag_seen) as data_last_ingestion_lag_seen, sum(data_eventcount) as data_eventcount by index,sourcetype | eval dcount_host=round(latest_dcount_host_5m, 2)
```

by:

```
| stats sum(eventcount_5m) as latest_eventcount_5m, avg(eventcount_5m) as avg_eventcount_5m, stdev(eventcount_5m) as stdev_eventcount_5m, perc95(eventcount_5m) as perc95_eventcount_5m, latest(latency_5m) as latest_latency_5m, avg(latency_5m) as avg_latency_5m, stdev(latency_5m) as stdev_latency_5m, perc95(latency_5m) as perc95_latency_5m, latest(dcount_host_5m) as latest_dcount_host_5m, avg(dcount_host_5m) as avg_dcount_host_5m, stdev(dcount_host_5m) as stdev_dcount_host_5m, perc95(dcount_host_5m) as perc95_dcount_host_5m, max(data_last_ingest) as data_last_ingest, min(data_first_time_seen) as data_first_time_seen, max(data_last_time_seen) as data_last_time_seen, avg(data_last_ingestion_lag_seen) as data_last_ingestion_lag_seen, sum(data_eventcount) as data_eventcount by index,sourcetype | eval dcount_host=round(latest_dcount_host_5m, 2)
```

**Comment by sebwurl** [comment]
No problem, hard to explain and to follow the issue :)

The change to "sum" will output the expected eventcount. I expect it to be close to the results of a raw search (if there is no significant ingest lag) and not like in my example from the first post with this huge gap. 

I'm not very familiar with the creation of ML models, but using the latest sounds very reasonable to me.

**Comment by guilhemmarchand** [comment]
Addressed in 2.0.25 which is going to be published today @sebwurl 

<img width="1751" alt="image" src="https://user-images.githubusercontent.com/7503059/233604095-d8d595c5-6247-47f2-95af-ba8652d8fff1.png">

**Comment by guilhemmarchand** [comment]
Thank you very much for your help on this one @sebwurl much appreciated!
